libbpf-tools/memleak.c

```diff
@@ -0,0 +1,707 @@
+	list[i + 1] = list[high];
+	list[high] = temp;
+	return (i + 1);
+}
+
+static void quick_sort(struct alloc_stack *list, int low, int high)
+{
+	int pi;
+
+	if (list == NULL) return;
+
+	if (low < high) {
+		pi = partition(list, low, high);
+		quick_sort(list, low, pi - 1);
+		quick_sort(list, pi + 1, high);
+	}
+}
+
+static void print_combined_outstanding_allocations(struct ksyms *ksyms, struct syms_cache *syms_cache,
+		struct memleak_bpf *obj)
+{
+	uint64_t stack_id = 0, next_stack_id;
+	const struct ksym *ksym;
+	const struct syms *syms;
+	const struct sym *sym;
+	int err, i, ifd, sfd, cnt;
+	unsigned long *ip;
+	time_t now;
+	struct tm * timeinfo;
+	struct combined_alloc_info_t info;
+	struct alloc_stack *stacks;
+	int max_prints = env.top;
+
+	time (&now);
+	timeinfo = localtime(&now);
+	printf("[%02x:%02d:%02d] Top %d stacks with outstanding allocations:\n",
+		timeinfo->tm_hour, timeinfo->tm_min, timeinfo->tm_sec, env.top);
+
+	ip = calloc(env.max_stack_depth, sizeof(*ip));
+	if (!ip) {
+		fprintf(stderr, "failed to alloccate memory of stack trace\n");
+		return;
+	}
+
+	stacks = calloc(MAX_HASH_ENTRY_NUM, sizeof(*stacks));
+	if (stacks == NULL) {
+		fprintf(stderr, "failed to allocate memory for stack info\n");
+		goto cleanup;
+	}
+	memset(stacks, 0, MAX_HASH_ENTRY_NUM * sizeof(*stacks));
+
+	cnt = 0;
+	ifd = bpf_map__fd(obj->maps.combined_allocs);
+	sfd = bpf_map__fd(obj->maps.stack_traces);
+	while (!bpf_map_get_next_key(ifd, &stack_id, &next_stack_id)) {
+		err = bpf_map_lookup_elem(ifd, &next_stack_id, &info);
+		if (err < 0) {
+			fprintf(stderr, "failed to lookup info: %d\n", err);
+			goto cleanup;
+		}
+		stack_id = next_stack_id;
+
+		stacks[cnt].stack_id = next_stack_id;
+		stacks[cnt].size = info.total_size;
+		stacks[cnt].nr = info.number_of_allocs;
+		cnt++;
+	}
+
+	quick_sort(stacks, 0, cnt);
+	if (cnt < max_prints) {
+		max_prints = cnt;
+	}
+
+	for (i=0; i < max_prints; i++) {
+		if (stacks[i].stack_id == 0) break;
+
+		if (bpf_map_lookup_elem(sfd, &stacks[i].stack_id, ip) != 0) {
+			fprintf(stderr, "    [Missed User Stack]\n");
+			continue;
+		}
+
+		if (env.kernel_trace) {
+			printf("\t%u bytes in %u allocations from stack\n", stacks[i].size, stacks[i].nr);
+			for (int j = 0; j < env.max_stack_depth && ip[j]; j++) {
+				ksym = ksyms__map_addr(ksyms, ip[j]);
+				if (ksym)
+					printf("\t\t%s (0x%lx)\n", ksym->name, ksym->addr);
+				else
+					printf("\t\t[unknown]\n");
+			}
+
+		} else {
+			syms = syms_cache__get_syms(syms_cache, env.pid);
+			if (!syms) {
+				fprintf(stderr, "failed to get syms\n");
+				continue;
+			}
+
+			printf("\t%u bytes in %u allocations from stack\n", stacks[i].size, stacks[i].nr);
+			for (int j = 0; j < env.max_stack_depth && ip[j]; j++) {
+				sym = syms__map_addr(syms, ip[j]);
+				if (sym)
+					printf("\t\t%s (0x%lx)\n", sym->name, sym->start);
+				else
+					printf("\t\t[unknown]\n");
+			}
+		}
+	}
+
+cleanup:
+	if (ip) free(ip);
+	if (stacks) free(stacks);
```

Comment:
free in reverse order ?

Reply:
Fix in commit [2beb289e](https://github.com/iovisor/bcc/pull/3828/commits/2beb289e98d651bddc2127299c5627b5d6149fce).